### DISCUSSION TITLE
vscode extension support

### DISCUSSION THREAD

**Discussion by howarddo2208** [zed-industries/zed#19158] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Hello Zed team,

As a user of Zed, I have been impressed with its speed and user-friendly interface. However, I believe that to attract more users to your platform, it would be beneficial to consider supporting VS Code extensions.

Many developers rely on VS Code extensions to enhance their workflow. The ability to leverage the vast library of VS Code extensions could make Zed more appealing to a wider audience. This could help improve the adoption of Zed. Moreover, I don't think many developers would want to rewrite their extensions for a small and new app like Zed, at least not just yet.
 
You can implement your custom extension API for more performance too, but it can come later. VS Code extension support is a must right out of the box imo.

**Comment by shinebayar-g** [comment]
Is it even possible?

**Comment by berkus** [comment]
> Is it even possible?

It is, but the amount of work here is pretty non-trivial:

- expose editor bindings for all the things that exist in vscode
  - bridge/thunk things that work differently
- support a JS runtime environment for plugins
- support compatible plugin configuration storage

**Comment by aelkoussy** [comment]
+1 with a suggestion
If you manage to add vscode extensions support with a better way of handling the really heavy extensions (like eslint for example), I think a lot of people would be excited to try and maybe switch from vscode to zed

So far, vscode is very fast and solid overall, with that problem above as a small nuisance - not sure if the host editor (vscode/zed) can control and improve the behaviour of extensions or not though tbh

**Comment by marek-nereca** [comment]
+1

**Comment by berkus** [comment]
I wonder if [boa](https://boajs.dev/blog/2024/03/07/boa-release-18) could become a JS engine to power the extensions.

**Reply by pnodet** [reply]
For reference, the Biome community and maintainers had a discussion about js engines here : https://github.com/biomejs/biome/discussions/2467

**Comment by kerns** [comment]
I would love to see this. I am sure it isn't trivial to implement, but the competitive adv. and stickiness of VS Code's extension ecosystem isn't trivial either.

**Comment by eimfach** [comment]
I think it would not make sense to support every vscode extension out there.. only a specific subset

**Comment by b2storefront** [comment]
can we at least short list some?

**Comment by subhasish-smiles** [comment]
This would be phenomenal !!
Was just thinking the same thing

**Comment by eimfach** [comment]
> can we at least short list some?

I think you need a specification on which parts of the VSCode Extension API to support as a start ..
And a way of telling if a specific extension does meet the supported criteria. Not an easy task I might say.

**Comment by himat** [comment]
Likewise, I don't think Zed has a shot of being a daily driver for a large part of the market until extensions are supported. I had finally switched from pure vim to vscode after many years only because there were certain AI extensions that made my workflow far superior. And I think zed would need those too.

**Comment by ShalokShalom** [comment]
> Likewise, I don't think Zed has a shot of being a daily driver for a large part of the market until extensions are supported. I had finally switched from pure vim to vscode after many years only because there were certain AI extensions that made my workflow far superior. And I think zed would need those too.

Zed now has AI features implemented.
See the homepage.

**Comment by thendodev** [comment]
I would like to say it doesn't matter how non-trivial the implementation is... the only real way to compete with VS Code is to support all if not most of their extensions otherwise, this is just another cool editor I can't use, there's no getting around that, it should probably be the most important focus for zed, that and the UI.

**Comment by berkus** [comment]
> I would like to say it doesn't matter how non-trivial the implementation is... the only real way to compete with VS Code is to support all if not most of their extensions

The good ones shall be reasonably easy to replicate. All the other duplicate crap probably is not very necessary.

I am personally just one extension short of the "comfortable set" and rather have fast and secure extensions in wasm rather than "do what you want with entire computer" JS from VSCode.

**Comment by ShalokShalom** [comment]
I imagine it makes sense to simply document the API in a way, that assists people with porting their apps.

Idk how far wasm extensions are supported.

Ideally, its possible to simply port existing extensions from one API to another - JS can be transpiled into wasm.

**Comment by notpeter** [comment]
While this is an interesting subject to consider, but at this time we have no plans to attempt a re-implementation the [VSCode Extension API](https://code.visualstudio.com/api) inside Zed as would be required to support VSCode extensions, and as a result I'm going to convert this to a github discussion.  Thanks for reporting.

**Comment by massone99** [comment]
It would be incredible to have this feature!

**Comment by naiyerasif** [comment]
Alternatively, a transpiler to easily convert a VSCode extension to Zed plugin until a developer is ready to support Zed would also be handy. This can be useful to reduce the initial migration effort for an extension developer.

**Reply by G2G2G2G** [reply]
yep this is the most possible route, especially with LLMs now

**Comment by vipexv** [comment]
I would suggest adding support for windows first.

**Reply by fredricknkum** [reply]
There is a Windows build but is not officially supported https://github.com/zed-industries/zed/discussions/15159#discussioncomment-10864307

**Reply by vipexv** [reply]
It's mostly broken for me, i'm saying a  stable build that is supported, as Windows is a big market and i think focusing on  that rather than adding extension support from VSCode would be better, as i personally don't use VSCode anymore.

I use neovim btw :]

**Reply by ShalokShalom** [reply]
I guess you ruled out the Linux runtime that is build into Windows?

**Reply by G2G2G2G** [reply]
> It's mostly broken for me, i'm saying a stable build that is supported, as Windows is a big market and i think focusing on that rather than adding extension support from VSCode would be better, as i personally don't use VSCode anymore.
> 
> I use neovim btw :]

No, windows users, like the general public, produce and do nothing of value for the large majority so there's no reason.

**Comment by thataboy** [comment]
zed needs a more extensive extension API first. Right now, the API is limited to LSP, themes, and slash commands.

If you build it, they will come. Or they may come. But you have to build it first.

**Comment by habruening** [comment]
Wow! 110 people like that idea.
I don't think, that it would ever be possible. Maybe it is technically feasible. But that is so much work. Have anybody ever done such a thing - reimplementing such a huge API? It is maybe not even a good API. It is just not realistic, I think. I guess the 110 people are mostly users and not programmers.
And I think, it would not even be a good idea. There are millions of extensions. But we don't need 30 extensions for bookmarks or 40 extensions for XML to be a good text editor. We must first be a good text editor. That alone attracts plugin developers. We want good and performant plugins and not just everything.

**Reply by pascalbehmenburg** [reply]
I think you are right. The past has shown that if an editor is good, people are willing to contribute. See emacs, nvim, vscode or even knowledge databases like logseq and obsidian. Might be best not to take on that burden of maintenance.

**Reply by ShalokShalom** [reply]
> There are millions of extensions. But we don't need 30 extensions for bookmarks or 40 extensions for XML to be a good text editor. We must first be a good text editor.

Some might say, there are already a million good text editors.

I suggest to look into the most popular extensions, and see if we lack any API, that prevents porting, or even the existence of such a kind of extension at all. 

Its probably enough to support the 100 most prominent extensions.

**Reply by berkus** [reply]
Since Zed has got it's own API now, it makes total sense to develop that and port over the necessary plugins from the VSCode ecosystem, I do not see a single reason to try and split the effort into supporting more than one plugin interface.

**Reply by james-lawrence** [reply]
the oni/oni2 editors did it, you might want to take a look/talk to that developer.

**Reply by JonGretar** [reply]
> the oni/oni2 editors did it, you might want to take a look/talk to that developer.

I think the Zed devs are simply not interested in that. And rightfully so. VSCode's plugin system is it's biggest weakness.

**Reply by james-lawrence** [reply]
@JonGretar simply pointing to prior art and feasibility oni/oni2 was a single dev iirc. *technically* very feasible.

**Reply by JonGretar** [reply]
> @JonGretar simply pointing to prior art and feasibility oni/oni2 was a single dev iirc. _technically_ very feasible.

Oh yes. Technically I agree. I'm more saying that it's a bad idea. 

A lot of VSCodes problems of being resource hungry is because if it's extensions. The editor itself is actually not **that** hungry. The amount of times my memory, or battery has died because some VSCode extension is leaking memory or stuck in 100% cpu loop is ridiculous. And usually when I did not bother to take the charger with me. 

Now of course writing the extensions in Rust is in no way a guarantee against memory leaks or other problems. But it does seem like JS apps tend to be more likely to have problems in this area. Especially as too many of them were written by web devs (such as myself) and we just tend not to have experience with dealing with long running apps/plugins.

**Comment by Cheirnomus** [comment]
One of the things I like most about Zed is that there are not a gazillion plugins. The best and most useful aspects that are typically plugins in VSCode are already baked into Zed. If Zed can continue that trend that would be great.

However, the existing Zed plugin system does need to be expanded and this is an excepted fact that is being worked on I'm sure. For instance, I'd like to see a Quokka plugin and Wallaby has informed me that they are waiting for the day when they can start developing for Zed.

**Comment by moheladwy** [comment]
I think, they might want to encourage the extensions developers to make a zed version of their extensions by making zed has a very well easy to use, documented, and supported extensions API.

However, the idea of having the speed and power of zed with the VS code extensions is magnificent.

**Reply by ShalokShalom** [reply]
I think that idea with making zed easy to extend, has not really shown to work yet, and that's not really a surprise. Everything is a tradeoff.

**Reply by berkus** [reply]
I'm fine with having zed-specific plugins if they are compiled into wasm and written in rust - this makes a much more robust foundation for the future plugins.

**Reply by ShalokShalom** [reply]
Wasm has the benefit of hosting different languages. 

This benefit is also mentioned on the homepage.

I question the sensibility of limiting that to Rust, considering that the amount of extensions is already almost zero. 

I also dont know, how this is supposed to increase the robustness of future plugins, or even those plugins itself.

How is Haskell, OCaml, Grain, Go, Ruby, or any of the other languages that handle memory safely, going to be "less robust than Rust"

https://github.com/Speykious/cve-rs

**Reply by berkus** [reply]
Exactly, you can write plugins in any other language that can be compiled to wasm.

Perhaps somebody will need to develop a bindings library for that though, and since it's a moving target currently it might be hard/unwise to do so - but nothing stops people from doing it.

**Reply by ShalokShalom** [reply]
Bindings can be created automatically, so long as we use wasmtime for the runtime.

See [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen), and [webassembly component model](https://component-model.bytecodealliance.org/)

**Comment by Karag0** [comment]
Too hard for bad feuture. Rust application and ts extension, bad idea

**Reply by james-lawrence** [reply]
ts extensions are not really the issue you just drop the ts code into the wasi runtime. all the benefits none of the problems.

**Comment by 0xSero** [comment]
I want to put my 2 cents here, given I have been working with Zed alongside Cursor, and VSCode. 

I have to switch back to vscode based IDEs often because of the extension support, might just be a skill issue on my end, but as a user I enjoy Zed more, I just want to be able to do my workflow with as little back and forth as possible. 

Long term it's best to get devs to build their extensions natively for Zed, the incentive to do so only comes from a bunch of people adopting this as their daily driver. 

Would Zed team put out a bounty ticket for someone to fork the repo, and figure out how to get a demo of this working? If so I would love to take it on myself.

**Reply by habruening** [reply]
I think, we must be a realistic. Making VS Code extension working for Zed is a very big job. I would not even bet a single cent on this idea. The world is not full of motivated Open Source developers for this idea. I think, the Zed team would not even accept the idea if you already had a perfect proof of concept. Not because it is a bad idea, but because it is just not realistic that we manage to ever get this working well enough.

**Comment by securesvet** [comment]
May one could provide any tips on how I, the VS Code extension developer, may get my extension working on Zed the easiest way with no pain? I think there might be people who already did so and I would love to learn from their experience. Thank you in advance

**Reply by Karag0** [reply]
No way. It needs full rewrite

**Comment by Ricardo-DEPT** [comment]
Is this discussion dead? Maybe more feasible now with agentic engineering? It's still blocking me from being able to switch completely sadly :(

**Reply by haroldao** [reply]
same...

**Reply by JonGretar** [reply]
Yes. The discussion is dead. VSCode extensions will not be supported in Zed at any point. 
Nor should they.

**Reply by habruening** [reply]
You don’t have to switch. Nobody forces you. VS Code is an extremely good text editor.

**Comment by Amiya167** [comment]
This doesn't seem like an easy task. First, it requires introducing Deno or other JavaScript runtimes, and implementing isolation and capability declarations to avoid security issues (although most VS Code plugins don't actually declare them, and some plugins rely on certain bugs or non-standard implementations; it's impossible to simulate even the bugs). Furthermore, for plugins that rely on WebView... adding a WebView framework (WebKitGTK on Linux might introduce numerous platform-specific bugs and could potentially degrade performance) is necessary.
